core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -287,69 +306,74 @@ public synchronized Optional<RemoteTask> scheduleTask(
             int partition,
             Multimap<PlanNodeId, Split> initialSplits)
     {
-        if (stateMachine.getState().isDone()) {
-            return Optional.empty();
-        }
+        try {
+            allTasksLock.writeLock().lock();
```

Comment:
>  if we use allTasksLock, do we still need synchronized on this method?

yes, we need because we use `allTasksLock` to synchronize accesses only to the `allTasks` set.

> allTasksLock
it is not true. There is an access in